blueprints/client-certificate-auth.md

```diff
@@ -0,0 +1,240 @@
+should be tested end-to-end with certificates as part of the integration tests.
+
+### Performance Impact
+
+There should be no serious performance impact of this beyond that being used
+for TLS itself. The basic costs of TLS are very slightly more than unencrypted
+transport, but this feature is entirely unavailable without TLS.
+
+### Security Impact
+
+The security impact of this feature will depend greatly on the operational
+practices of the operator. Private root keys need to be properly maintained and
+certificates should be rotated frequently. If a key or intermediate is
+compromised, the root should be rotated and all keys reissued immediately.
+
+Root certificates are global and do not apply to tenants in particular. Any
+root cert can provide proof of identity for any user of any tenant.
+
+Someone with administrative credentials could easily create a backdoor with a
+long-lived cert that allows them to authenticate as any user. Operators will
+need to keep tight control of their private keys. This is not worse than the
+existing situation, though, because someone with administrative access to a
+Traffic Ops system already has such access.
+
+If Traffic Ops is serving unencrypted traffic, this feature will not
+meaningfully improve security.
+
+### Upgrade Impact
+
+An upgraded system will initially contain an empty list of authorized roots.
+This means that no client certificates will be valid and since invalid
+certificates are ignored, the login behaviour will be identical.
+
+#### Downgrade Impact
+
+A system that is downgraded will lose the ability to understand client
+certificates. Any clients that were relying on them will naturally lose their
+ability to authenticate.
+
+### Operations Impact
+
+There is no operational impact for those who do not wish to use Client Certificates.
+
+Using Client Certificates, however, requires an operator to have a robust and
+secure Public Key Infrastructure. This is a significant endeavour that will
+require operators to either manually manage a considerable number of keys and
+certificates or to develop tooling that manages it automatically.
+
+The PKI properly belongs entirely outside of Traffic Control, however, because
+operators are likely to have their own requirements and existing tooling around
+how this is handled.
+
+If an operator uses a caching proxy to access Traffic Ops, it will need to be
+issued an empty-UID client certificate in order to proxy authentication. It
+will also need to validate presented certs against the authorized roots, copy
+the Subject into the `Client-Cert-Subject` header, add the Forwarded header,
+and include the UID in the cache key.
+
+### Developer Impact
+
+Developers with test systems may find it's easier to develop on test systems
+using certificates instead of passwords. It will present an additional option
+at least. Existing workflows will work just fine, though.
+
+### Alternatives
+
```

Comment:
I've only recently become aware of it, but is [SQRL](https://www.grc.com/sqrl/sqrl.htm) an alternative as well?

Reply:
SQRL relies on a consistent private key, similar to mTLS. While it could form the basis of a protocol, everything already supports mTLS and very little supports SQRL. Likewise, SQRL doesn't have protection for things like replay, expiration, third-party attestation and delegation the way mTLS does. This is fine, because SQRL is trying to solve a very different problem. It's not really a solid option here, though. I guess I could add a section on SQRL, but it doesn't really offer what we're looking for here.